enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,516 @@
+* Resilient storage - see future enhancement
+* Support for platforms other than bare metal including automated ci testing
+* Support for other topologies (eg. hypershift)
+* Adding worker nodes
+* Creation RHEL-HA events and metrics for consumption by the OpenShift monitoring stack (Deferred to post-MVP)
+
+## Proposal
+
+Use the RHEL-HA stack (Corosync, and Pacemaker), which has been used to delivered supported 2-node cluster experiences for multiple decades, to manage cri-o, kubelet, and the etcd daemon.
+etcd will run as as a voting member on both nodes.
+We will take advantage of RHEL-HA's native support for systemd and re-use the standard cri-o and kublet units, as well as create a new Open Cluster Framework (OCF) script for etcd.
+The existing startup order of cri-o, then kubelet, then etcd will be preserved.
+The `etcdctl`, `etcd-metrics`, and `etcd-readyz` containers will remain part of the static pod.
+
+Use RedFish compatible Baseboard Management Controllers (BMCs) as our primary mechanism to power off (fence) an unreachable peer and ensure that it can do no harm while the remaining node continues.
+
+Upon a peer failure, the RHEL-HA components on the surivor will fence the peer and use the OCF script to restart etcd as a new cluster-of-one.
+
+Upon a network failure, the RHEL-HA components ensure that exactly one node will survive, fence it's peer, and use the OCF script to restart etcd as a new cluster-of-one.
+
+In both cases, the control-plane's dependance on etcd will cause it to respond with errors until etcd has been restarted.
+
+Upon rebooting, the RHEL-HA components ensure that a node remains inert (not running cri-o, kubelet, or etcd) until it sees it's peer.
+If the peer is likely to remain offline for an extended period of time, admin confirmation is required to allow the node to start OpenShift.
+
+When starting etcd, the OCF script will use etcd's cluster ID and version counter to determine whether the existing data directory can be reused, or must be erased before joining an active peer.
+
+### Workflow Description
+
+#### Cluster Creation
+
+Creation of a two node control-plane will be possible via the core installer (with an additional bootstrap node), and via the Assisted Installer (without an additional bootstrap node).
+
+In the case of the core OpenShift installer, the user-facing proceedure is unchanged from a standard "IPI" installation, other than the configuration of 2 nodes instead of 3.
+Internally, the RedFish details for each node will need to make their way into the RHEL-HA configuration, but this is information already required for bare-metal hosts.
+
+In the case of the Assisted Installer, the user-facing proceedure follows the standard flow except for the configuration of 2 nodes instead of 3, and the collection of RedFish details for each node which are needed for the RHEL-HA configuration.
```

Comment:
ZTP/ABI is the end goal, but IIUC we need to pass through SaaS to get there.
For now, I have highlighted the fact that passwords are involved.

Reply:
There's no reason to ever put this in the UI imho.
For ZTP it could be implemented without being part of the ReST API, because the controllers for the CRs are directly connected to the internals.
For ABI it would have to be part of the ReST API, although we could use authz to disable it in the SaaS.

Reply:
ZTP supports disconnected, but requires a hub cluster. It's part of MCE, so you don't need all of ACM though.

ABI's niche is installing your first cluster in an environment (i.e. no hub available), particularly if either (a) it is disconnected, or (b) this is a process you want to repeat many times in similar environments, or (c) you have existing automation you want to use for booting the hosts.